[feature][processing] "Array of Offset Features" algorithm

nyalldawson opened the pull request:
This algorithm creates copies of features in a layer, by creating multiple offset versions of the feature. Each copy is displaced by a preset amount in the x/y/z/m axis.

nyalldawson committed:
[feature][processing] "Array of Offset Features" algorithm

This algorithm creates copies of features in a layer, by
creating multiple offset versions of the feature. Each copy is displaced
by a preset amount in the x/y/z/m axis.

nirvn commented on src/analysis/processing/qgsalgorithmarrayfeatures.cpp:
```diff
@@ -0,0 +1,221 @@
+
+QString QgsArrayFeaturesAlgorithm::outputName() const
+{
+  return QObject::tr( "Offset" );
```
"Offset" is also the output name for the offset line algorithm's output. I'd suggest "Offsets Array", or maybe just add an "s"?

nirvn commented:
100% cool, 0% practical use of this algorithm:
![image](https://user-images.githubusercontent.com/1728657/43253006-18669c26-90ee-11e8-8fc8-a6aa3cfa7d93.png)

:)

This'll come in handy, esp. with all parameters being dynamic.

nyalldawson committed:
Tweak output name

nyalldawson committed:
[FEATURE][processing] "Create parallel lines" algorithm

This algorithm creates copies of line features in a layer, by
creating multiple parallel versions of each feature. Each copy is offset
by a preset distance.

nyalldawson committed:
[processing][feature] Port 'offset lines' to c++, support dynamic offset parameter

Adds data defined support for the offset line distance parameter.

nirvn commented:
Thanks for the name tweaking, looking good.

nyalldawson commented:
Bonus features: a new "create parallel lines" algorithm, and a port of "offset lines" to c++ (with new dynamic distance parameter)
